### PR TITLE
Improve launch time of WebKit processes

### PR DESCRIPTION
**Pull request by pvollan**
#### c61ab1f570586f0ceaab0f4eb690ea1718940eca
<pre>
Improve launch time of WebKit processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=260198">https://bugs.webkit.org/show_bug.cgi?id=260198</a>
rdar://113901020

Reviewed by Brent Fulgham.

Experiments show that it is signficantly faster to launch WebKit processes as extensions with ExtensionKit.
This patch is working towards this by creating empty executables for these extensions and starting
adopting ExtensionKit (<a href="https://developer.apple.com/documentation/extensionkit)">https://developer.apple.com/documentation/extensionkit)</a> as the long-term architecture
for WebKit auxiliary processes as a lighter-weight mechanism for launching and handling processes.

This patch is written by Alex Christensen and Per Arne Vollan.

* Configurations/CommonBase.xcconfig:
* Source/WebKit/Configurations/BaseExtension.xcconfig: Added.
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/GPUExtension.xcconfig: Added.
* Source/WebKit/Configurations/NetworkingExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCaptivePortalExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentCrashyExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentExtension.xcconfig: Added.
* Source/WebKit/Shared/API/Cocoa/WKMain.h:
* Source/WebKit/Shared/API/Cocoa/WKMain.mm:
(WKExtensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.swift: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm: Added.
(extensionMain):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/NetworkingExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-CaptivePortal-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Crashy-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Info.plist: Added.
* Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268228@main">https://commits.webkit.org/268228@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c9374784f9546703090d0a6a8f93a2bd96d021a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19045 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19997 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17804 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19554 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19559 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16569 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21793 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16576 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23746 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17617 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21693 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17190 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21547 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17936 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -17706,15 +18110,22 @@
...
 		};
 		1A50DB3A110A3C13000D3FE5 /* Release */ = {
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = A1EDD2D91884ACE000BBFE98 /* All.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
 				PRODUCT_NAME = All;
+				SWIFT_OBJC_BRIDGING_HEADER = "Shared/AuxiliaryProcessExtensions/Framework, XPC Services, Extensions, and daemons-Bridging-Header.h";
+				SWIFT_VERSION = 5.0;
 			};
```
Comment (emw-apple) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Be sure to remove these, and put all settings overrides in xcconfigs. Xcode adds some settings here by default when you create a new target, but we don't want them.

Ditto for the other `XCBuildConfiguration` objects below this.

File: Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "AuxiliaryProcessExtensionBridge.h"
+
+#import <WebKit/WKMain.h>
+
+void extensionMain(int argc, char** argv)
+{
+    WKExtensionMain(argc, argv);
```
Comment (emw-apple) [inline Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:32]: nit: we should probably handle the return value here, right?
```suggestion
    exit(WKExtensionMain(argc, argv));
```

File: Source/WebKit/Configurations/BaseExtension.xcconfig [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+#include "BaseTarget.xcconfig"
+
+CLANG_ENABLE_OBJC_ARC = YES;
+INSTALL_PATH = $(INSTALL_PATH_PREFIX)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(SYSTEM_LIBRARY_DIR)/ExtensionKit/Extensions/;
+CODE_SIGN_ENTITLEMENTS = Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtension.entitlements;
+SWIFT_OBJC_BRIDGING_HEADER = Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.h;
+PRODUCT_NAME = $(TARGET_NAME);
+CLANG_USE_OPTIMIZATION_PROFILE = NO;
+CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING_ENABLE_LLVM_PROFILE_GENERATION = NO;
```
Comment (emw-apple) [inline Source/WebKit/Configurations/BaseExtension.xcconfig]: I assume the intent is to disable PGO instrumentation entirely? You can just set `CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO`.

**Review by emw-apple — COMMENTED** [review]
Nice! Let me know if I can be of help testing this.

**Comment by weinig** [comment]
Out of interest, what in ExtensionKit makes process launching faster? My understanding is that under the covers there would also be an xpc service.

File: Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+    <key>com.apple.webkit.webcontent.extension</key>
+    <dict>
+        <key>EXExtensionPointIsPublic</key>
+        <true/>
```
Comment (xeenon) [inline Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt]: I think you want all these `EXExtensionPointIsPublic` values to be `<false/>`. These are not extension points that third-party apps can implement extensions for.

**Comment by xeenon** [comment]
> What in ExtensionKit makes process launching faster? My understanding is that under the covers there would also be an xpc service.
> 
> Given that ExtensionKit is written on top of what WebKit uses, its not clear to me how it could be lighter-weight.

That is my understanding as well @weinig — it is just a wrapper around XPC. I am surprised by this too.

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj (hunk 1 above)
Comment (pvollan) [reply Source/WebKit/WebKit.xcodeproj/project.pbxproj]: Fixed!

Thanks for reviewing!

File: Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm (hunk 2 above)
Comment (pvollan) [reply Source/WebKit/Shared/AuxiliaryProcessExtensions/AuxiliaryProcessExtensionBridge.mm:32]: Good point, I have not addressed this yet, but will do.

File: Source/WebKit/Configurations/BaseExtension.xcconfig (hunk 3 above)
Comment (pvollan) [reply Source/WebKit/Configurations/BaseExtension.xcconfig]: Fixed.

**Comment by pvollan** [comment]
> What in ExtensionKit makes process launching faster? My understanding is that under the covers there would also be an xpc service.
> 
> Given that ExtensionKit is written on top of what WebKit uses, its not clear to me how it could be lighter-weight.

That is a good question. TBH, I am not sure how this is implemented under the hood, but we have measured significant launch time improvements with this approach.

File: Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt (hunk 4 above)
Comment (pvollan) [reply Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt]: Fixed!

Thanks for reviewing, weinig, xeenon, and mew-apple!

**Comment by weinig** [comment]
> > What in ExtensionKit makes process launching faster? My understanding is that under the covers there would also be an xpc service.
> > Given that ExtensionKit is written on top of what WebKit uses, its not clear to me how it could be lighter-weight.
> 
> That is a good question. TBH, I am not sure how this is implemented under the hood, but we have measured significant launch time improvements with this approach.

Perhaps it goes without saying, but I think it is pretty crucial to understand this before committing to the approach.

File: Configurations/CommonBase.xcconfig [hunk 5]
```diff
@@ -55,6 +55,13 @@ OTHER_LDFLAGS = $(inherited) $(WK_COMMON_OTHER_LDFLAGS);
 WK_COMMON_OTHER_TAPI_FLAGS = -x objective-c++ -std=c++2a -fno-rtti $(WK_SANITIZER_OTHER_TAPI_FLAGS);
 OTHER_TAPI_FLAGS = $(inherited) $(WK_COMMON_OTHER_TAPI_FLAGS);
 
+WK_DEFAULT_SWIFT_OPTIMIZATION_LEVEL = -O;
+WK_DEFAULT_SWIFT_OPTIMIZATION_LEVEL[config=Debug] = -Onone;
+SWIFT_OPTIMIZATION_LEVEL = $(WK_DEFAULT_SWIFT_OPTIMIZATION_LEVEL);
+
+WK_DEFAULT_SWIFT_VERSION = 5.0;
+SWIFT_VERSION = $(WK_DEFAULT_SWIFT_VERSION);
```
Comment (brentfulgham) [inline Configurations/CommonBase.xcconfig]: I'm a bit surprised that these rules aren't visible to `BaseTarget.xcconfig` or `BaseExtension.xcconfig`. I wonder if we should have a Swift xcconfig file (not necessary for this patch).

**Review by brentfulgham — APPROVED** [review]
Assuming you incorporate Elliott's suggested change, I think this looks great.

File: Configurations/CommonBase.xcconfig (hunk 5 above)
Comment (pvollan) [reply Configurations/CommonBase.xcconfig]: That is a good point. I think the changes here may not be required. Will fix.

Thanks for reviewing!

**Comment by pvollan** [comment]
> Assuming you incorporate Elliott's suggested change, I think this looks great.

I have a followup-patch almost ready that addresses the issue where the return value of WKExtensionMain is not handled. Would you be ok with handling this in a follow-up patch?

Thanks for reviewing!

**Pull request merged**